### DISCUSSION TITLE
Include a stable CVE-like field in advisory data

### DISCUSSION THREAD

**Discussion by G-Rath** [npm/feedback#604] [Registry]
(I've opened this as being for the Registry, but let me know if there's a better category/place for me to put this request)

In the npm v6 days, the advisory data we'd get from `npm audit --json` would have a lot of good info:

```
{
  "findings": [
    {
      "version": "1.12.1",
      "paths": ["jquery-ui"]
    }
  ],
  "metadata": null,
  "vulnerable_versions": "<1.13.0",
  "module_name": "jquery-ui",
  "severity": "moderate",
  "github_advisory_id": "GHSA-j7qv-pgf6-hvh4",
  "cves": ["CVE-2021-41183"],
  "access": "public",
  "patched_versions": ">=1.13.0",
  "updated": "2021-10-25T22:06:46.000Z",
  "recommendation": "Upgrade to version 1.13.0 or later",
  "cwe": "CWE-79",
  "found_by": null,
  "deleted": null,
  "id": 1006727,
  "references": "- https://github.com/jquery/jquery-ui/security/advisories/GHSA-j7qv-pgf6-hvh4\n- https://nvd.nist.gov/vuln/detail/CVE-2021-41183\n- https://github.com/jquery/jquery-ui/pull/1953\n- https://blog.jqueryui.com/2021/10/jquery-ui-1-13-0-released/\n- https://bugs.jqueryui.com/ticket/15284\n- https://security.netapp.com/advisory/ntap-20211118-0004/\n- https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/NXIUUBRVLA4E7G7MMIKCEN75YN7UFERW/\n- https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/O74SXYY7RGXREQDQUDQD4BPJ4QQTD2XQ/\n- https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/SNXA7XRKGINWSUIPIZ6ZBCTV6N3KSHES/\n- https://github.com/advisories/GHSA-j7qv-pgf6-hvh4",
  "created": "2021-11-22T19:00:44.964Z",
  "reported_by": null,
  "title": "XSS in `*Text` options of the Datepicker widget in jquery-ui",
  "npm_advisory_id": null,
  "overview": "### Impact\nAccepting the value of various `*Text` options of the Datepicker widget from untrusted sources may execute untrusted code. For example, initializing the datepicker in the following way:\n```js\n$( \"#datepicker\" ).datepicker( {\n\tshowButtonPanel: true,\n\tshowOn: \"both\",\n\tcloseText: \"<script>doEvilThing( 'closeText XSS' )</script>\",\n\tcurrentText: \"<script>doEvilThing( 'currentText XSS' )</script>\",\n\tprevText: \"<script>doEvilThing( 'prevText XSS' )</script>\",\n\tnextText: \"<script>doEvilThing( 'nextText XSS' )</script>\",\n\tbuttonText: \"<script>doEvilThing( 'buttonText XSS' )</script>\",\n\tappendText: \"<script>doEvilThing( 'appendText XSS' )</script>\",\n} );\n```\nwill call `doEvilThing` with 6 different parameters coming from all `*Text` options.\n\n### Patches\nThe issue is fixed in jQuery UI 1.13.0. The values passed to various `*Text` options are now always treated as pure text, not HTML.\n\n### Workarounds\nA workaround is to not accept the value of the `*Text` options from untrusted sources.\n\n### For more information\nIf you have any questions or comments about this advisory, search for a relevant issue in [the jQuery UI repo](https://github.com/jquery/jquery-ui/issues). If you don't find an answer, open a new issue.",
  "url": "https://github.com/advisories/GHSA-j7qv-pgf6-hvh4"
}
```

This included fields like `cves` and `cwe` - in fact TIL it now has `github_advisory_id` which is awesome.

With npm v7 we got a new auditing endpoint used by `npm audit --json` that doesn't include nearly as much information:

```
{
  "name": "jquery-ui",
  "severity": "moderate",
  "via": [
    {
      "source": 1006727,
      "name": "jquery-ui",
      "dependency": "jquery-ui",
      "title": "XSS in `*Text` options of the Datepicker widget in jquery-ui",
      "url": "https://github.com/advisories/GHSA-j7qv-pgf6-hvh4",
      "severity": "moderate",
      "range": "<1.13.0"
    }
  ],
  "effects": [],
  "range": "<1.13.0",
  "nodes": ["node_modules/jquery-ui"],
  "fixAvailable": true
}
```

This was workable however as the `source` field seems to have the `id` value from the old format - this allowed [tools to be built](https://github.com/G-Rath/audit-app) for ignoring vulnerabilities based on their advisory id & path.

However, recently we've seen those ids change - initially they changed when the npm advisory database was merged into the Github advisory database (all the IDs seem to have been prefixed with `100`), which we figured would be a one-off event and so weren't fussed about, but we've since seen IDs change unexpected.

The latest was yesterday/today, with three vulnerabilities in `jquery-ui` - I generated an `.auditapprc.json` with ignores for `1006720|jquery-ui`, `1006721|jquery-ui`, and `1006722|jquery-ui`, and today they've changed to 1006727, 1006728, and 1006729 respectively.

I'm hoping this is just some residual noise from the advisory database merging, and so that it'll stop after a bit of time, but I think it would be good if the new audit endpoint could include one or more fields with a stable (enough) ID like what was included in the old endpoint.

(it does look like the urls are stable and present for every advisory, so in theory I could parse the GH advisory ID from that, but it'd be nice to have a dedicated field for this so I don't have to worry about that url structure changing or what characters to allow when extracting, etc)

**Comment by G-Rath** [comment]
I noticed while gathering details for this that the new audit endpoint doesn't seem to be including all of the advisories: auditing our app with the old endpoint reports the three jQuery UI vulnerabilities mentioned but the new endpoint only reports the one.

I'm really hoping this is some delayed effect but I find it concerning given that the advisories were published 27 days ago and the patched version (1.13.0) was published 2 months ago, so I'd expect the checker to be accurate at this point (e.g. I could understand if it was a case of them being brand new advisories that needed to propagated through the system because caching and whatnot).

Happy to make a new discussion or issue somewhere for this, but want to confirm where the best place to do so is first :)

**Reply by G-Rath** [reply]
Ok this looks like it's just a bug in the cli, as the report does include the three advisories:

```
npm sill audit report   'jquery-ui': [
npm sill audit report     {
npm sill audit report       id: 1006727,
npm sill audit report       url: 'https://github.com/advisories/GHSA-j7qv-pgf6-hvh4',
npm sill audit report       title: 'XSS in `*Text` options of the Datepicker widget in jquery-ui',
npm sill audit report       severity: 'moderate',
npm sill audit report       vulnerable_versions: '<1.13.0'
npm sill audit report     },
npm sill audit report     {
npm sill audit report       id: 1006728,
npm sill audit report       url: 'https://github.com/advisories/GHSA-gpqq-952q-5327',
npm sill audit report       title: 'XSS in the `of` option of the `.position()` util in jquery-ui',
npm sill audit report       severity: 'moderate',
npm sill audit report       vulnerable_versions: '<1.13.0'
npm sill audit report     },
npm sill audit report     {
npm sill audit report       id: 1006729,
npm sill audit report       url: 'https://github.com/advisories/GHSA-9gj3-hwp5-pmwc',
npm sill audit report       title: 'XSS in the `altField` option of the Datepicker widget in jquery-ui',
npm sill audit report       severity: 'moderate',
npm sill audit report       vulnerable_versions: '<1.13.0'
npm sill audit report     }
npm sill audit report   ],
```

**Reply by G-Rath** [reply]
Have reported as https://github.com/npm/arborist/issues/361

**Comment by MylesBorins** [comment]
hey @G-Rath the non-audit related advisory fetch endpoints are deprecated and we are advising folks to use the GitHub Advisory DB for the exact reasons you've identified in this discussion. Please let me know if you have additional questions.

Did you notice the deprecation header when you were doing your get requests?

https://github.blog/changelog/2021-11-16-deprecating-non-audit-related-advisory-fetch-endpoints-for-the-npmjs-com-registry-api/

**Reply by G-Rath** [reply]
@MylesBorins I'm not using any endpoint directly, this is using `npm audit --json`.
